### ISSUE TITLE
requireAuth redirecting to login even when user is logged in

### ISSUE THREAD

**Issue by erikwithuhk** [erikwithuhk/mason-jar-salads#2]
## Context

In my app, I've passed in utils/requireAuth.js to the App component via onEnter, and my requireAuth.js script checks if the user is logged in with the expression `firebase.auth().currentUser === null`, and redirects them to the welcome screen if they are not logged in.

However, this redirect is occurring even when a user is logged in.
## What have you tried?

I've isolated the problem to the `firebase.auth().currentUser === null` expression in my utils/requireAuth module. If I log the response of `firebase.auth().currentUser` when a user is logged in, it logs `null`, even though the object returned when I run `firebase.auth()` contains an object `.currentUser`. So the expression is returning `null` whether or not a user is logged in.

I've read through the Firebase docs and also compared it with the code that we used for the user auth lesson on Tuesday (and it's exactly the same as far as I can tell). I also tried [implementing a solution I found on Google Groups](https://groups.google.com/forum/?hl=vi#!topic/firebase-talk/836OyVNd_Yg), but it did not resolve the problem.

I also tried putting a setTimeout on my requireAuth module for 200ms, and while this resolved the problem of not redirecting to the welcome page when a user is signed in, it prevented the redirect to the welcome page when the user wasn't logged in.
## Question

Is there a reason that my `firebase.auth().currentUser === null` expression in my requireAuth module would be returning true when a user is logged in?

**Comment by amadden80** [comment]
``` js
firebase.auth().onAuthStateChanged((currentuser) => {
  if (currentuser) {
    console.log('USER:', currentuser);
  } else {
    console.log('USER NOT SIGNED IN');
  }
});
```

Looks like the `firebase.auth()` return allows for a `onAuthStateChanged` callback.  Cool!  
With that, you can get the user WHEN it is ready.

**Comment by amadden80** [comment]
@WilliamfShaw  Have you seen this `onAuthStateChanged` listener?  Seems this could help with this firebase.auth() async issue.

**Comment by erikwithuhk** [comment]
It's working! Want to share with all of you because it seems to be just a little weird React Router quirk.

I used the `onAuthStateChanged` listener Andrew suggested, but the redirect was still not functioning properly.

When David and I were working on the problem last night, we found in the [React Router docs](https://github.com/ReactTraining/react-router/blob/master/docs/API.md#onenternextstate-replace-callback) that "If callback is listed as a 3rd argument [of React Router `onEnter`], this hook will run asynchronously, and the transition will block until callback is called."

I thought this meant that the callback would have to be defined and passed in (e.g. running `firebase.auth()`), but then I found [this post](https://github.com/gaearon/react-hot-loader/issues/329) and [this post](https://github.com/ReactTraining/react-router/issues/3671) that just call `callback()`, so I assume this means that `onEnter` passes a callback function in which then gets called.

Here's how my working `requireAuth` looks:

```
import React, { Component } from 'react';
import firebase from '../../firebase.config.js';

function requireAuth(nextState, replace, callback) {
  firebase.auth().onAuthStateChanged((user) => {
    if (user === null) {
      replace({
        pathname: '/welcome',
        state: { nextPathname: nextState.location.pathname },
      });
    }
    callback();
  });
}

module.exports = requireAuth;
```
